build-tools/fetch-windows-assemblers/fetch-windows-assemblers.cs

```diff
@@ -0,0 +1,697 @@
+			destFile.Write (buffer, 0, bread);
+			fread += bread;
+			dread += (uint)bread;
+		}
+
+		destFile.Flush ();
+		destFile.Close ();
+		destFile.Dispose ();
+		Extract (destFileName, lfh.CRC32);
+
+		if (dread != lfh.CompressedSize)
+			Console.Error.WriteLine ($"  Invalid data size: expected {lfh.CompressedSize} bytes, read {dread} bytes");
+
+		return true;
+	}
+
+	static void Extract (string compressedFilePath, uint crc32FromHeader)
+	{
+		string outputFile = Path.GetFileNameWithoutExtension (compressedFilePath);
+		using (var fs = File.OpenRead (compressedFilePath)) {
+			using (var dfs = File.OpenWrite (outputFile)) {
+				Extract (fs, dfs, crc32FromHeader);
+			}
+		}
+	}
+
+	static void Extract (Stream src, Stream dest, uint crc32FromHeader)
+	{
+		uint fileCRC = 0;
+		int fread = 0;
+		var crc32 = new CRC32 ();
+		var buffer = new byte [8192];
+
+		using (var iis = new DeflateStream (src, CompressionMode.Decompress)) {
+			while (true) {
+				fread = iis.Read(buffer, 0, buffer.Length);
+				if (fread <= 0)
+					break;
+
+				fileCRC = crc32.Append (fileCRC, buffer, 0, fread);
+				dest.Write (buffer, 0, fread);
+			}
+			dest.Flush ();
+		}
+
+		if (fileCRC != crc32FromHeader)
+			Console.Error.WriteLine ($"  Invalid CRC32: expected 0x{crc32FromHeader:x}, got 0x{fileCRC:x}");
+	}
+
+	static async Task<(bool success, Stream data)> ReadFileData (HttpClient httpClient, Uri url, CDHeader cdh)
+	{
+		long fileOffset = cdh.RelativeOffsetOfLocalHeader;
+		long dataSize =
+			cdh.CompressedSize +
+			30 + // local file header size, the static portion
+			cdh.FileName.Length + // They're the same in both haders
+			cdh.ExtraFieldLength + // This may differ between headers...
+			16384; // ...so we add some extra padding
+
+		var req = new HttpRequestMessage (HttpMethod.Get, url);
+		req.Headers.ConnectionClose = true;
+		req.Headers.Range = new RangeHeaderValue (fileOffset, fileOffset + dataSize);
+
+		HttpResponseMessage resp = await httpClient.SendAsync (req).ConfigureAwait (false);
+
+		if (!resp.IsSuccessStatusCode) {
+			Console.Error.WriteLine ($"Failed to read file data: HTTP error {resp.StatusCode}");
+			return (false, null);
+		}
+
+		Stream s = await resp.Content.ReadAsStreamAsync ().ConfigureAwait (false);
+		if (s.Length < dataSize) {
+			Console.Error.WriteLine ($"Failed to read file data: invalid data length ({s.Length} < {dataSize})");
+			s.Dispose ();
+			return (false, null);
+		}
+
+		return (true, s);
+	}
+
+	static (bool success, LFHeader lfh) ReadLFHeader (BinaryReader cdr, long dataLength, ref long nread)
+	{
+		var lfh = new LFHeader ();
+
+		bool worked;
+		lfh.Signature = ReadUInt (cdr, dataLength, ref nread, out worked);
+		if (!worked || lfh.Signature != LFHeaderSignature) {
+			Console.Error.WriteLine ($"Invalid signature ({lfh.Signature:x} != {LFHeaderSignature:x})");
+			goto failed;
+		}
+
+		lfh.VersionNeededToExtract = ReadUShort (cdr, dataLength, ref nread, out worked);
+		if (!worked) {
+			goto failed;
```

Comment:
Will we need a way to "remotely" (via build logs) determine *why* one of these failed?  We have 12 possibly different reasons for something to fail, but no way of knowing which one.

This is (arguably) a place where exceptions are better, in that the exception will capture the line number for the `throw`...